server/aesdsocket.c

```diff
@@ -18,320 +11,548 @@
+                free(tmp);
+                tmp = NULL;
+                count_nodes1++;
+                printf("nodes deletes from threads %d\n",count_nodes);
+                break;    
+            }
+        }
+        usleep(10);
+    }       
+    return NULL;        
+}
+
+
+void *timestamp_func(void *timestamp)
+{
+    while(1)
+    {
+       char outstr[200] ={0};
+           time_t t;
+           struct tm *tmp;
+           
+           t = time(NULL);
+           tmp = localtime(&t);
+           if (tmp == NULL) {
+               perror("localtime");
+           }
+
+           if (strftime(outstr, sizeof(outstr),"timestamp: %a %d %b %Y %T", tmp) == 0) 
+           {
+               perror("strftime error");
+               syslog(LOG_ERR,"strftime error");
+               //exit(//exit_FAILURE);
+           }
+           
+           strcat(outstr,"\n");
+
+           printf("Result string is %s\n", outstr);
+           
+       
+        pthread_mutex_lock(&mutex);
+        //open file to write the received packets in append mode
+        fd = open(path, O_RDWR | O_CREAT | O_APPEND, 0777);
+        if(fd < 0)
+        {
+            printf("File not created for timestamp\n");
+        }
+        else
+        {
+            printf("File created successfully in append mode for timestamp\n");
+        }   
 
+        //write packets in the file opened
+        int stampret = write(fd, outstr, strlen(outstr));
+        if(stampret == -1)
+        {
+            perror("write failed");
+            syslog(LOG_ERR,"write failed");
+        }
+        close(fd);
+        pthread_mutex_unlock(&mutex);   
 
+        sleep(10);
+    }
+
+}
+
+    
 //handler for SIGINT and SIGTERM
 static void signal_handler(int signo)
 {
-	if((signo == SIGINT) || (signo == SIGTERM))
-	{
-		syslog(LOG_DEBUG,"Caught SIGNAL, exiting");
-		
-		
-		if(shutdown(socketfd, SHUT_RDWR) == -1)
-		{
-			perror("Signal Handler : Shutdown");
-			syslog(LOG_ERR,"Shutdown failure");
-			closelog();
-			exit(EXIT_FAILURE);
-		}
-		
-		if(socketfd)
-		{
-			close(socketfd);
-		}
-		if(newfd)
-		{
-			close(newfd);
-		}
-		if(fd)
-		{
-			close(fd);
-		}
-		
-		if(remove(path)==0)
-		{
-			syslog(LOG_DEBUG,"File removed successfully");
-		}
-		else
-		{
-			syslog(LOG_DEBUG,"File not deleted");
-			closelog();
-			exit(EXIT_FAILURE);
-		}
-	}
-	exit(EXIT_SUCCESS);
+        syslog(LOG_DEBUG,"Caught SIGNAL, //exiting");
+        
+        printf("Caught SIGNAL, //exiting\n");
+        
+        node_t *handle = NULL;
+        TAILQ_FOREACH(handle, &head, nodes)
+        {
+            if((shutdown(handle->client_fd, SHUT_RDWR)) == -1)
+            {
+                syslog(LOG_ERR,"Shutdown failure from threads");
+            }
+            
+            if((pthread_kill(handle->ids, SIGKILL))!=0)
+            {
+                syslog(LOG_ERR,"Unable to kill thread");
+            }
+            
+        }
+        
+        _free_queue(&head);
+        printf("Hi\n");
```

Comment:
@DivyeshShashikant is it safe to call linkedlist APIs, printf etc from signal handler?

Reply:
> @DivyeshShashikant is it safe to wave Hi here? 😜 is it safe to call linkedlist APIs, printf etc from signal handler?(async signal unsafe funcs)

Oh i completely forgot to remove the debug print